### ISSUE TITLE
please expose intent classification results for all all intent classifiers in pipeline

### ISSUE THREAD

**Issue by lisongshan** [RasaHQ/rasa#1035]
**Rasa NLU version**:0.12.2

**Operating system** (windows, osx, ...):all

**Content of model configuration file**:
```yml

```

**Issue**:
There is a strange discrepancy between entity recognition and intent classification:  message object   hold all entities recognized by different extractors , but it only hold one result by current intent classifier. So there is no way for a custom component to (model) assemble results from different classifiers. Please add intent list to message class and have each classifier to add their classification result to the list.

**Comment by tmbo** [comment]
I mean the reason is a kind of that the idea is that a message can only have one intent and multiple entities. But of course that is debatable, especially the intent one. What is your use case for using multiple intent classifiers?

**Comment by lisongshan** [comment]
I think you missed my question. A message has one intent doesn't mean different classifiers will give the same intent classification. So this request is more about what you do when different classifiers give different intent classification. A standard practice is to use model assemble. Different classifiers may give different intent classifications, but I would like to use a custom component to decide which one is the best. Another good use case would be: say I have a  binary classifier for "unknown" and "known" intent (using only in-domain training data), then I would need to write a custom component to combine the regular classifier with this binary classifier. Then the custom component need to access classification results by previous two classifiers.

I understand that there is a stronger reason for multiple entity recognizers because different recognizer can recognizer different entities. But supporting model assemble is a good reason as well.

In summary, this request is to allow developers to do model assemble. And the change is small, just add to message object a container to store  the intents and intent rankings from different classifiers.

**Comment by akelad** [comment]
Hmm, but using multiple intent classifiers would be a bit confusing now? Say one gave you the most probable intent to be `affirm` with probability 0.8 and the other classifier gave you the intent `deny` with probability 0.95. You can't really directly compare those probabilities because they're different classifiers and therefore will predict these probabilities differently.

**Comment by lisongshan** [comment]
good examples! if you can only use one classifier,  you are forced to  believe one or the other. But in many cases, it would be more prudent to take both into account, for example, take average of them (with the help of a custom component). More concrete, say one classifier gives intent_ranking:{intent1:0.6, intent2:0.4}, another classifier gives intent_ranking:{intent1:0.2, intent2:0.8}, then the average would be as intent_ranking:{intent1:0.4, intent2:0.6}. This approach is called model assemble which gives better accuracy in most of the cases.

**Comment by benjamin-work** [comment]
Not sure how easy it is to plug into Rasa NLU, but for the described case, there are solutions from sklearn (e.g. [BaggingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html), [VotingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html)) that can wrap multiple classifiers and aggregate the results. The ensemble just behaves like a normal classifier would, so that the user doesn't need to adjust the call.

**Comment by amn41** [comment]
I think on balance it does not make sense to do this in Rasa NLU. I can certainly see the case for using a model ensemble, but that should be a custom component (e.g. a `EnsembleModelIntentClassifier`) which can be added to your pipeline without modifying NLU at all.

**Comment by lisongshan** [comment]
@amn41 
If you take a closer look about how built-in intent classifier set its classification result,  you will find that currently, the message object only stores intent classifications for one intent classifier. An example will make it clear: say we configure the model with intent_classifier_sklearn, intent_classifier_tensorflow_embedding and a custom_intent_classifier, in that order. The problem is that the later custom_intent_classifier can not get the classification result from intent_classifier_sklearn, because the way built-in classifiers were implemented is to overwrite previous classification.

 My Request of enhancement is to each classifier add classification result to message instead of overwriting it. As I noted in my original request, the built-in entity recognizers just add their entities to message, not overwriting. I was surprised why classifier and entity recognizer behave differently in this regard.

**Comment by lisongshan** [comment]
@disimone 
please see my comments above. You may take a look the code to verify it.

**Comment by deepak209g** [comment]
Any updates on this issue ? I just started exploring rasa and this feature is something that would help me.

**Comment by mohammedterry** [comment]
Its strange that when training Rasa Core's Policy it allows for an ensemble of classifiers to be used.   It would make sense to do the same for Rasa NLU's Intent classifiers.

**Comment by wochinge** [comment]
@Ghostvv @tabergma @dakshvar22 @JEM-Mosig What are your thoughts on this one?

**Comment by Ghostvv** [comment]
@wochinge we discussed it several times, it is not clear how to compare outputs from different classifiers

**Comment by wochinge** [comment]
Ok, then I close this issue for now.
